File: src/agent/telemetry.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import { nanoid } from '@/agent/utils/nanoid';
import type { RouteDecision } from '@/agent/types';

const STORAGE_KEY = 'riflett_traces_v2';
const MAX_TRACES = 100;

let storageLock = Promise.resolve<void>(undefined);

const withStorageLock = async <T>(task: () => Promise<T>): Promise<T> => {
  let release: (() => void) | null = null;
  const next = new Promise<void>((resolve) => {
    release = resolve;
  });
  const previous = storageLock;
  storageLock = previous.then(() => next);
  try {
    await previous;
    return await task();
  } finally {
    release?.();
  }
};
export interface RetrievalTelemetry {
  id: string;
  kind: string;
  compositeScore: number;
  scoring: {
    recency: number;
    priority: number;
    semantic: number;
    affect: number;
    relationship: number;
  };
}

export interface PlannerTelemetry {
  action: string | null;
  ask: string | null;
  payloadPreview: string | null;
}

export type ActionStatus = 'pending' | 'accepted' | 'declined' | 'failed' | 'none';

export interface ActionTelemetry {
  type: string;
  status: ActionStatus;
  ids: string[];
  metadata?: Record<string, unknown>;
}

export interface ConfidenceTelemetry {
  retrieval: number;
  plan: number;
  overall: number;
}

export interface TraceEvent {
  id: string;
  ts: number;
  maskedUserText: string;
  intentLabel: string;
  intentConfidence: number;
  decision: RouteDecision;
  retrieval: RetrievalTelemetry[];
  redactionSummary: Record<string, number>;
  latencyMs?: number;
  planner?: PlannerTelemetry | null;
  action?: ActionTelemetry | null;
  confidence?: ConfidenceTelemetry | null;
  receipts?: string[];
}

const loadTraces = async (): Promise<TraceEvent[]> => {
  try {
    const raw = await AsyncStorage.getItem(STORAGE_KEY);
    if (!raw) return [];
    const parsed = JSON.parse(raw) as TraceEvent[];
    if (!Array.isArray(parsed)) return [];
    return parsed;
  } catch {
    return [];
  }
};

const saveTraces = async (traces: TraceEvent[]): Promise<void> => {
  const trimmed = traces
    .sort((a, b) => b.ts - a.ts)
    .slice(0, MAX_TRACES);
  try {
    await AsyncStorage.setItem(STORAGE_KEY, JSON.stringify(trimmed));
  } catch (error) {
    console.warn('[telemetry] save failed', error);
  }
};

export const Telemetry = {
  async record(params: {
    maskedUserText: string;
    intentLabel: string;
    intentConfidence: number;
    decision: RouteDecision;
    retrieval: RetrievalTelemetry[];
    redactionSummary: Record<string, number>;
    startedAt: number;
  }): Promise<string> {
    return withStorageLock(async () => {
      const traces = await loadTraces();
      const id = nanoid();
      traces.unshift({
        id,
        ts: Date.now(),
        maskedUserText: params.maskedUserText,
        intentLabel: params.intentLabel,
        intentConfidence: params.intentConfidence,
        decision: params.decision,
        retrieval: params.retrieval,
        redactionSummary: params.redactionSummary,
        latencyMs: Date.now() - params.startedAt,
        planner: null,
        action: null,
      });
      await saveTraces(traces);
      return id;
    });
  },

  async update(id: string, patch: Partial<TraceEvent>): Promise<void> {
    await withStorageLock(async () => {
      const traces = await loadTraces();
      const next = traces.map((trace) =>
        trace.id === id
          ? {
              ...trace,
              ...patch,
            }
          : trace
      );
      await saveTraces(next);
    });
  },
};

File: services/python-ai/app/agent/telemetry.py
import asyncio
import time
from typing import Any, Dict, List, Optional

from .types import RouteDecision
from .utils.nanoid import nanoid

STORAGE_KEY = "riflett_traces_v2"
MAX_TRACES = 100

_storage_lock = asyncio.Lock()
_trace_store: List[Dict[str, Any]] = []


async def _with_storage_lock(task):
    async with _storage_lock:
        return await task()


async def _load_traces() -> List[Dict[str, Any]]:
    return list(_trace_store)


async def _save_traces(traces: List[Dict[str, Any]]) -> None:
    trimmed = sorted(traces, key=lambda trace: trace["ts"], reverse=True)[:MAX_TRACES]
    _trace_store.clear()
    _trace_store.extend(trimmed)


class Telemetry:
    @staticmethod
    async def record(params: Dict[str, Any]) -> str:
        async def _task():
            traces = await _load_traces()
            trace_id = nanoid()
            traces.insert(
                0,
                {
                    "id": trace_id,
                    "ts": int(time.time() * 1000),
                    "maskedUserText": params["maskedUserText"],
                    "intentLabel": params["intentLabel"],
                    "intentConfidence": params["intentConfidence"],
                    "decision": params["decision"],
                    "retrieval": params.get("retrieval", []),
                    "redactionSummary": params.get("redactionSummary", {}),
                    "latencyMs": int(time.time() * 1000) - params.get("startedAt", int(time.time() * 1000)),
                    "planner": None,
                    "action": None,
                },
            )
            await _save_traces(traces)
            return trace_id

        return await _with_storage_lock(_task)

    @staticmethod
    async def update(trace_id: str, patch: Dict[str, Any]) -> None:
        async def _task():
            traces = await _load_traces()
            next_traces: List[Dict[str, Any]] = []
            for trace in traces:
                if trace.get("id") == trace_id:
                    merged = {**trace, **patch}
                    next_traces.append(merged)
                else:
                    next_traces.append(trace)
            await _save_traces(next_traces)

        await _with_storage_lock(_task)

